pkg/apis/storage/validation/validation.go

```diff
@@ -266,3 +266,178 @@ func validateAllowedTopologies(topologies []api.TopologySelectorTerm, fldPath *f
+	return allErrs
+}
+
+// ValidateCSIDriverInfoSpecs tests that the specified CSIDriverInfoSpecs
+// have valid data.
+func validateCSIDriverInfoSpecs(
+	drivers []storage.CSIDriverInfoSpec, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if drivers == nil || len(drivers) == 0 {
+		return allErrs
+	}
+
+	for i, driver := range drivers {
+		idxPath := fldPath.Index(i)
+		allErrs = append(allErrs, validateCSIDriverInfoSpecName(driver.Name, idxPath.Child("name"))...)
+		allErrs = append(allErrs, validateCSIDriverInfoSpecNodeID(driver.NodeID, idxPath.Child("nodeID"))...)
+		allErrs = append(allErrs, validateCSIDriverInfoSpecTopologyKeys(driver.TopologyKeys, idxPath.Child("topologyKeys"))...)
+	}
+
+	return allErrs
+}
+
+// ValidateCSIDriverInfoStatuses tests that the specified CSIDriverInfoStatuses
+// have valid data.
+func validateCSIDriverInfoStatuses(
+	drivers []storage.CSIDriverInfoStatus, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if drivers == nil || len(drivers) == 0 {
+		return allErrs
+	}
+
+	for i, driver := range drivers {
+		idxPath := fldPath.Index(i)
+		allErrs = append(allErrs, validateCSIDriverInfoStatusName(driver.Name, idxPath.Child("name"))...)
+		allErrs = append(allErrs, validateCSIDriverInfoStatusVolumePluginMechanism(string(driver.VolumePluginMechanism), idxPath.Child("volumePluginMechanism"))...)
+	}
+
+	return allErrs
+}
+
+// validateCSIDriverInfoSpecName tests if Name in CSIDriverInfoSpec is a valid name.
+func validateCSIDriverInfoSpecName(name string, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if len(name) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath, name))
+	}
+	return allErrs
+}
+
+// validateCSIDriverInfoStatusName tests if Name in CSIDriverInfoStatus is a valid name.
+func validateCSIDriverInfoStatusName(name string, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if len(name) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath, name))
+	}
+	return allErrs
+}
+
+// validateCSIDriverInfoSpecNodeID tests if Name in CSIDriverInfoSpec is a valid node id.
+func validateCSIDriverInfoSpecNodeID(nodeID string, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if len(nodeID) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath, nodeID))
+	}
+	return allErrs
+}
+
+// validateCSIDriverInfoSpecTopologyKeys tests if TopologyKeys in CSIDriverInfoSpec are
+// valid topology keys.
+func validateCSIDriverInfoSpecTopologyKeys(keys []string, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if len(keys) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath, ""))
+	}
+	for _, key := range keys {
+		if len(key) == 0 {
+			allErrs = append(allErrs, field.Required(fldPath, key))
+		}
+	}
+	return allErrs
+}
+
+var supportedVolumePluginMechanisms = sets.NewString(string("in-tree"), string("csi"))
```

Comment:
nit: global var declarations at top of file in one big block

Reply:
Done